cohttp-lwt/src/client.ml

```diff
@@ -12,33 +14,348 @@ module Make (IO : S.IO) (Net : S.Net with module IO = IO) = struct
+        uri
+      =
+      let mutex = Lwt_mutex.create () in
+      Lwt.async (fun () -> Lwt_mutex.lock mutex);
+      let channels =
+        Net.connect_uri ~ctx uri >|= fun (_, ic, oc) ->
+        (ic, oc)
+      in
+      let connection =
+        { finalise
+        ; in_flight = Queue.create ()
+        ; conn = Connecting channels
+        ; mutex
+        ; condition = Lwt_condition.create ()
+        }
+      in
+      Lwt.on_any channels
+        begin fun channels ->
+          connection.conn <- Full channels;
+          Lwt_mutex.unlock connection.mutex;
+          Lwt.async (fun () -> reader connection);
+        end
+        begin fun e ->
+          connection.conn <- Failed e;
+          Lwt_mutex.unlock connection.mutex;
+        end;
+      connection
+  end
+
+  module Conn_cache =
+  struct
+    type t = < request : ?body:Body.t -> Request.t -> (Response.t * Body.t) Lwt.t >
+    module Uri_map = Map.Make (Uri)
+
+    let no_cache ?ctx:ctx () :t = object
+      method request ?body req =
+        let connection = Connection.connect ?ctx (Request.uri req) in
+        let res = Connection.request connection ?body req in
+        (* this can be simplified when https://github.com/mirage/ocaml-conduit/pull/319 is released. *)
+        Lwt.async begin fun () ->
+          res >>= fun (_, body) ->
+          match body with
+          | `Empty | `String _ | `Strings _ -> Lwt.return_unit
+          | `Stream stream ->
+            Lwt_stream.closed stream >>= fun () ->
+            Connection.close connection
+        end;
+        res
+    end
+
+    (*
+     * Select from existing connections by hostname, not IP.
+     * One hostname may resolve to multiple IP addresses.
+     * In such a case we want only one connection.
+     * The downside is that some resolver redirections may break.
+     *)
+
+    let strip_uri uri =
+      Uri.(make ?scheme:(scheme uri) ?host:(host uri) ?port:(port uri) ())
+
+    let cache ?ctx:ctx ?(keep=60_000) ?(retry=2) ?(parallel=4) ?(depth=1) () =
+      object(self)
+        val mutable cache :Connection.t list Uri_map.t = Uri_map.empty
+        val mutex = Lwt_mutex.create ()
+
+        method private get_connection addr =
+          let finalise connection =
+            Lwt_mutex.with_lock mutex @@ fun () ->
+            let conns = Uri_map.find addr cache in
+            let conns = List.filter ((!=) connection) conns in
+            if conns = []
+            then cache <- cache |> Uri_map.remove addr
+            else cache <- cache |> Uri_map.add addr conns;
+            Lwt.return_unit
+          in
+          let connect () =
+            let connection = Connection.connect ~finalise ?ctx addr
+            and timeout = ref Lwt.return_unit in
+            let rec busy () =
+              Lwt.cancel !timeout;
+              if Connection.length connection = 0 then begin
+                timeout :=
+                  Time.sleep_ns Int64.(mul (of_int keep) 1_000_000L) >>= fun () ->
```

Comment:
Do we really need this sleep 0.001s call here?

Reply:
Would be nice if we can avoid threading through that. Otherwise we should restructure the code to add explicit timeouts to the calls, which could also solve https://github.com/mirage/ocaml-cohttp/issues/790

Reply:
This is a configureable timeout on idle connections. `keep` is the number of idle milliseconds after which a connection is closed.

Regarding #790 I also had a timeout in my `Connection` module, but removed it. I could add it again (but probably better in a separate pull request). The timeout needs to be tickled after every successful read of a response header or body stream chunk (via `Lwt_stream.map`).